src/codegen/expression.rs

```diff
@@ -2370,75 +2264,47 @@ pub fn emit_function_call(
-                                selector,
-                            ),
-                        );
-
-                        (
-                            encoder.abi_encode(loc, &args, ns, vartab, cfg).0,
-                            Some(address),
-                        )
-                    } else {
-                        tys.insert(0, Type::Address(false));
-                        tys.insert(1, Type::Address(false));
-                        tys.insert(2, Type::Uint(64));
-                        tys.insert(3, Type::Bytes(4));
-                        tys.insert(4, Type::Bytes(1));
-
-                        (
-                            Expression::AbiEncode {
-                                loc: *loc,
-                                tys,
-                                packed: vec![
-                                    address,
-                                    Expression::Builtin(
-                                        *loc,
-                                        vec![Type::Address(false)],
-                                        Builtin::GetAddress,
-                                        Vec::new(),
-                                    ),
-                                    value.clone(),
-                                    Expression::NumberLiteral(*loc, Type::Bytes(4), BigInt::zero()),
-                                    Expression::NumberLiteral(*loc, Type::Bytes(1), BigInt::zero()),
-                                    Expression::BytesLiteral(
-                                        *loc,
-                                        Type::Bytes(selector.len() as u8),
-                                        selector,
-                                    ),
-                                ],
-                                args,
-                            },
-                            None,
-                        )
-                    }
+                let address = if dest_func.selector.is_some()
+                    || matches!(ns.target, Target::Substrate { .. })
+                {
+                    args.insert(
+                        0,
+                        Expression::BytesLiteral(*loc, Type::Bytes(selector.len() as u8), selector),
+                    );
+                    Some(address)
                 } else {
-                    (
-                        Expression::AbiEncode {
-                            loc: *loc,
-                            tys,
-                            packed: vec![Expression::BytesLiteral(
-                                *loc,
-                                Type::Bytes(selector.len() as u8),
-                                selector,
-                            )],
-                            args,
-                        },
-                        Some(address),
-                    )
+                    let mut encoder_args: Vec<Expression> = Vec::with_capacity(6 + args.len());
```

Comment:
Maybe I'm wrong here, but this if is confusing to read. Shouldn't we explicitly match the target? I don't understand why we are differentiating here between  the Solana target vs. Substrate Target OR selector.is_some()... What happens when we add more targets?
```rust
let address = match {
    Target::Solana => { ... },
    Target::Substrate { .. } => { ... },
    _ => unimplemented!()
}
```

Reply:
I agree this piece of code here isn't the best one. The match would work well, but it would need repeated code as there one case that is exactly the same for both Solana and Substrate. According to @seanyoung , Solana's API for external call is about to change, so this ugliness will go away soon.

Reply:
Any thought on this @seanyoung ?

Reply:
The `if` condition would make more sense like so:
```rust
   if ns.target == Target::Solana && dest_func.selector.is_none() {
       // swap then and else statement block
   } 
``` 
This ugliness will disappear soon when we remove all the extra stuff from the instruction data on Solana.